src/CommonLib/LDAPUtils.cs

```diff
@@ -1415,163 +1438,534 @@ private static bool RequestNETBIOSNameFromComputer(string server, string domain,
-                connection.SessionOptions.Signing = false;
+                //Attempt an initial bind. If this fails, likely auth is invalid, or its not a valid target
+                connection.Bind();
+            }
+            catch (LdapException e)
+            {
+                connection.Dispose();
+                return new LdapConnectionTestResult(false, e, null, null);
             }
 
-            if (_ldapConfig.SSL)
-                connection.SessionOptions.SecureSocketLayer = true;
+            try
+            {
+                //Do an initial search request to get the rootDSE
+                //This ldap filter is equivalent to (objectclass=*)
+                var searchRequest = new SearchRequest("", new LDAPFilter().AddAllObjects().GetFilter(),
+                    SearchScope.Base, null);
+                searchRequest.Controls.Add(new SearchOptionsControl(SearchOption.DomainScope));
+
+                var response = (SearchResponse)connection.SendRequest(searchRequest);
+                if (response?.Entries == null)
+                {
+                    connection.Dispose();
+                    return new LdapConnectionTestResult(false, null, null, null);
+                }
 
-            if (_ldapConfig.DisableCertVerification)
-                connection.SessionOptions.VerifyServerCertificate = (ldapConnection, certificate) => true;
+                if (response.Entries.Count == 0)
+                {
+                    connection.Dispose();
+                    return new LdapConnectionTestResult(false, new LdapException((int)LdapErrorCodes.KerberosAuthType), null, null);
+                }
 
-            connection.AuthType = authType;
+                var entry = response.Entries[0];
+                var baseDN = entry.GetProperty(LDAPProperties.RootDomainNamingContext).ToUpper().Trim();
+                var configurationDN = entry.GetProperty(LDAPProperties.ConfigurationNamingContext).ToUpper().Trim();
+                var domainname = Helpers.DistinguishedNameToDomain(baseDN).ToUpper().Trim();
+                var servername = entry.GetProperty(LDAPProperties.ServerName);
+                var compName = servername.Substring(0, servername.IndexOf(',')).Substring(3).Trim();
+                var fullServerName = $"{compName}.{domainname}".ToUpper().Trim();
 
-            _ldapConnections.AddOrUpdate(targetServer, connection, (s, ldapConnection) =>
+                return new LdapConnectionTestResult(true, null, new DomainInfo
+                {
+                    DomainConfigurationPath = configurationDN,
+                    DomainSearchBase = baseDN,
+                    DomainFQDN = domainname
+                }, fullServerName);
+            }
+            catch (LdapException e)
             {
-                ldapConnection.Dispose();
-                return connection;
-            });
+                try
+                {
+                    connection.Dispose();
```

Comment:
No action needed here, just thinking about the logic and flow here. There are a lot of calls in this logic to `Dispose` - we may want to, in the future, consider inverting control using a closure to encapsulate the logic so that `Dispose` is called only once in the function that receives the closure.

```
public TestResult TestConnection(SomeFunc testFunc)  {
    open_connection = OpenConnection(...)

    try {
        testFunc(connection)
    } finally {
        open_connection.Dispose()
    }
}
```

Reply:
We're specifically disposing of the connection only in error cases here, so its not quite so straightforward as just wrapping it and disposing